### ISSUE TITLE
Allow plugins to indicate dependencies on random files... discovered during the transform

### ISSUE THREAD

**Issue by chancancode** [babel/babel#16016]
### 💻

- [ ] Would you like to work on this feature?

### What problem are you trying to solve?

This is a re-open of #8497, which was supposedly fixed by #14065.

In the original issue, it cited something like https://www.npmjs.com/package/babel-plugin-inline-import and other macro-like plugins as motivating use cases. However, as far as I can tell, what was eventually implemented in #14065 did not actually solve it. It's certainly possible that I am missing something though, perhaps @nicolo-ribaudo can chime in.

Consider a simplified version of `babel-plugin-inline-import`.

Let's call it `babel-plugin-include` – the feature added is an `include("some-file")` macro, which reads the referenced file and dumps its content inline as a string, [similar to the Rust version](https://doc.rust-lang.org/std/macro.include.html#examples):

```js
// my-app/src/censor.js

const BANNED_WORDS = include("./banned-words.txt").split("\n");

export default function censor(text) {
  // ...do something about the banned words in `text`
  return text;
}
```

For this to work, essentially, someone need to keep the metadata that:

1. When the source of `my-app/src/censor.js` changes, it needs to be recompiled (the usual)
2. In addition, when `my-app/src/banned-words.txt` changes, it also needs to be recompiled, even if `censor.js` itself did not change

On the surface, #14065 appears to be trying to solve this problem, but as I looked deeper, it does not appear to be the case. It solves a very specific version of this problem, like [react-native-dotenv](https://github.com/goatandsheep/react-native-dotenv) (@goatandsheep 👋🏼 thanks for writing the blog post on this feature), where the plugin depends on a fixed set of external dependencies know at plugin-instantiation time, independent of the contents of the code being transformed.

In the case of `react-native-dotenv`, @nicolo-ribaudo provided an example on how they could adopt this feature. However, I don't think the same solution would work for `babel-plugin-include` or many of the other ideas mentioned in the original issue – you would run into multiple problems:

1. The callbacks for `api.cache.using(...)` have no contextual information (e.g. the filename of the file being transformed)
2. You cannot call `api.addExternalDependency(...)` inside the visitor – it's too late, the array of external dependency has already been frozen and you'll get a runtime error

And actually, I am not sure that the implemented solution _really_ works for `react-native-dotenv` either. From their README, it seems like they actually depend on more than just the content of the physical `.env` file, like reading values from runtime-set environment variables, or switching between different `.env.*` files depending on a number of factors. It seems like, the only really correct "logical cache key" is only really know after transforming a specific file (e.g. a hash of all the consumed variable's values, please if any of the `.env.*` files changed).

So, it seems to me there is an impedance mismatch between what we set out wanting to solve and what was actually implemented. Ultimately, I'm not sure how much `@babel/cli` is "actually used", but the bigger goal seems to be to set up some kind of metadata infrastructure/pattern that other consumers like `babel-loader` can consume. Returning the list of external dependencies from `transform()` seems to accomplish this alright, but it's the part where who you can populate this list that is where the mismatch lies. The common pattern is to "instantiate" (it seems like core calls it "loadPluginDescriptor") plugins once and reuse them across transforms, and currently the api limits the cache key generation to the "instantiation" phase, which seems wrong to me. But then again I could totally be missing something.

### Describe the solution you'd like

The ability to influence the caching behavior on a "per-plugin, per-transformed-source-file (or source content)" basis, rather than just "per-plugin".

### Describe alternatives you've considered

Using some convoluted login inside the callbacks to `api.cache.using(...)` and friends, but ultimately I don't think it has enough information/context to do anything useful. Plus, not being able to modify the list of external dependencies at transform time is also a dealbreaker for these kind of use cases.

**Comment by JLHwung** [comment]
Thanks for the thorough write-up. It will be maintainers' dream world if every issue is like this. 😄

In webpack land, an alternative (and working now) solution for the use case mentioned above would be to make `./banned-words.txt` into a normal dependency, and load it via the `raw-loader`:
```js
// my-app/src/censor.js

import BANNED_WORDS from "./banned-words.txt";

export default function censor(text) {
  // ...do something about the banned words in `text`
  return text;
}
```

```js
// webpack.config.js
module.exports = {
  module: {
    rules: [
      {
        test: /\.txt$/i,
        use: 'raw-loader',
      },
    ],
  },
};
```

Here Babel has no knowledge about `./banned-words.txt` and the bundler will take care of the dependency watching.

The `addExternalDependency` was provided for those implicit dependency, hence the term "external", mostly tools config which are mostly not imported directly. For tools config, the assumption is that they are known at the compile time.

I acknowledge that `include` macro is just an example, yet the solution above also reveals that if we do have per-file "external" dependencies _and_ it is known at compile time. Then we can also make it a direct normal dependency and leverage the bundlers to do the watching, since even if we somehow registered them into external dependencies, Babel still has to pass through that information to the bundler anyway, as Babel core is not the watcher. By utilizing the standard import syntax, we get extra benefits like `eslint-plugin-import` that can lint if the `./banned-words.txt` are moved to somewhere but the import usage is not updated accordingly.

**Comment by chancancode** [comment]
Yes, I agree that instead of implementing `include()` as a macro, it could have been implemented natively within the bundler and the problems would go away.

In fact, I think you could even say that about @goatandsheep's `react-native-dotenv` – it uses the native import syntax anyway so instead of implementing it as a babel plugin you could have just synthesized the '@env' module the native bundler extension point and let it shake out the unneeded stuff, you loose the inlining but that is probably not that big of a deal in the grand scheme of things.

The `include()` example is just meant to be a simple illustrative example of what a big part of what (to me, at least) the original issue seemed to be about. 

I think one reason the babel solution is seems attractive is you can implement it once, and rely on the integration with the bundlers to do the right thing, rather than implement N versions, one for each bundler. At least that seems to be the desire to me – having the metadata in there, `@babel/core` itself is fairly agnostic about what they mean, but `babel-loader` etc can plumb that through to the host as it is appropriate.

I think the core of the problem is that the both the current caching semantics and the available metadata is too coarse – even in cases where you _are_ able to use `addExternalDependency` as it is implemented today, it is unfortunate that you are forced to invalidate _all_ files ever seen by the plugin – if you have a large code base with thousands of files and you use `react-native-dotenv`, whenever you change one ENV variable, you invalidate the cache for _all_ the files in your codebase, even if just a handful of them actually use the dotenv feature.

**Comment by chancancode** [comment]
In an ideal world, the plugin would register the external dependency only on the source file that uses the feature, or in `@babel/core` terminology, "the relevant `transform()` call. The infrastructure is half there, in that the list of external dependencies are returned per-`transform()` call, and `babel-loader` is wiring that up correctly on its end. The issue/impedance mismatch comes form the fact that there isn't a good way to actually _add_ that information "per-`transform()` call", due to the pattern of how plugins are actually instantiated outside of core.
